crypto/dh/dh_key.c

```diff
@@ -246,24 +246,36 @@ int dh_buf2key(DH *dh, const unsigned char *buf, size_t len)
     return 0;
 }
 
-size_t dh_key2buf(const DH *dh, unsigned char **pbuf)
+size_t dh_key2buf(const DH *dh, unsigned char **pbuf_out)
 {
+    unsigned char *pbuf;
+    const BIGNUM *p;
     const BIGNUM *pubkey;
-    int bn_size;
+    int pubkey_size;
+    int pad = 0;
 
+    DH_get0_pqg(dh, &p, NULL, NULL);
     DH_get0_key(dh, &pubkey, NULL);
-    if (pubkey == NULL || (bn_size = BN_num_bytes(pubkey)) == 0) {
+    if (p == NULL || pubkey == NULL
+            || (pubkey_size = BN_num_bytes(pubkey)) == 0) {
         DHerr(DH_F_DH_KEY2BUF, ERR_R_PASSED_INVALID_ARGUMENT);
         return 0;
     }
-    if ((*pbuf = OPENSSL_malloc(bn_size)) == NULL) {
+    if (pubkey_size < BN_num_bytes(p)) {
+        pad = BN_num_bytes(p) - pubkey_size;
+    }
+    if ((pbuf = OPENSSL_malloc(pad + pubkey_size)) == NULL) {
         DHerr(DH_F_DH_KEY2BUF, ERR_R_MALLOC_FAILURE);
         return 0;
     }
-    if (BN_bn2bin(pubkey, *pbuf) < 0) {
-        OPENSSL_free(*pbuf);
+    if (pad) {
+        memset(pbuf, 0, pad);
+    }
+    if (BN_bn2bin(pubkey, pbuf + pad) < 0) {
```

Comment:
Probably using `BN_bn2binpad()` is better, rather than doing the memset first.

Reply:
Changed. And Squashed.